### PR TITLE
distribution: errors: do not retry if no credentials provided

### PR DESCRIPTION
**Pull request by runcom**
Fixes case `c)` in #21054

/cc @aaronlehmann 

Signed-off-by: Antonio Murdaca runcom@redhat.com

### CODE REVIEW

**Comment by runcom** [comment]
This should probably differentiate between pull and push I guess

**Comment by aaronlehmann** [comment]
Thanks for the PR!

File: integration-cli/docker_cli_push_test.go [hunk 1]
```diff
@@ -526,3 +526,11 @@ func (s *DockerTrustSuite) TestTrustedPushWithReleasesDelegation(c *check.C) {
 	c.Assert(err, check.IsNil, check.Commentf("Unable to read targets/releases metadata"))
 	c.Assert(string(contents), checker.Contains, `"latest"`, check.Commentf(string(contents)))
 }
+
+func (s *DockerRegistryAuthSuite) TestPushNoCredentialsNoRetry(c *check.C) {
+	repoName := fmt.Sprintf("%s/busybox", privateRegistryURL)
+	dockerCmd(c, "tag", "busybox", repoName)
+	out, err := dockerCmdWithError("push", repoName)
+	c.Assert(err, check.NotNil, check.Commentf(out))
```
Comment (aaronlehmann) [inline integration-cli/docker_cli_push_test.go]: Should this check for a specific error message?

Maybe we should check the output to make sure it doesn't include `Retrying`?

**Comment by aaronlehmann** [comment]
Weird, CI didn't run on this PR.

ping @icecrime

**Comment by thaJeztah** [comment]
@aaronlehmann oh, that happens sometimes if GitHub misses a ping. Doing a `!rebuild` will help, let me do that

**Comment by thaJeztah** [comment]
Yup, worked, they're here now :-)

File: integration-cli/docker_cli_push_test.go (hunk 1 above)
Comment (runcom) [inline integration-cli/docker_cli_push_test.go]: I'm having troubles running tests on my laptop so I did a test PR while fixing this test

**Comment by runcom** [comment]
Btw I think this fix will regress on the pull side, it won't return a 404 anymore when credentials aren't found I suppose (need to check)

**Comment by runcom** [comment]
This PR is also adding tests for case a) and d) - we'll reiterate after we *fix* everything in #21054

**Comment by runcom** [comment]
@aaronlehmann this is good to review if all tests pass :)

File: integration-cli/docker_cli_push_test.go [hunk 2]
```diff
@@ -526,3 +526,21 @@ func (s *DockerTrustSuite) TestTrustedPushWithReleasesDelegation(c *check.C) {
...
 	c.Assert(string(contents), checker.Contains, `"latest"`, check.Commentf(string(contents)))
 }
+
+func (s *DockerRegistryAuthSuite) TestPushNoCredentialsNoRetry(c *check.C) {
+	repoName := fmt.Sprintf("%s/busybox", privateRegistryURL)
+	dockerCmd(c, "tag", "busybox", repoName)
+	out, _, err := dockerCmdWithError("push", repoName)
+	c.Assert(err, check.NotNil, check.Commentf(out))
+	c.Assert(out, check.Not(checker.Contains), "Retrying")
+	c.Assert(out, checker.Contains, "no basic auth credentials")
```
Comment (runcom) [inline integration-cli/docker_cli_push_test.go]: I'm wondering if the error string here should be the same as the docker.io case below

File: integration-cli/docker_cli_push_test.go [hunk 3]
```diff
@@ -526,3 +526,21 @@ func (s *DockerTrustSuite) TestTrustedPushWithReleasesDelegation(c *check.C) {
...
+}
+
+// This may be flaky but it's needed not to regress on unauthorized push, see #21054
+func (s *DockerSuite) TestPushToCentralRegistryUnauthorized(c *check.C) {
+	repoName := "test/busybox"
```
Comment (runcom) [inline integration-cli/docker_cli_push_test.go]: This needs testrequires network (I'll add it tomorrow)

**Comment by calavera** [comment]
LGTM

**Comment by aaronlehmann** [comment]
LGTM

**Pull request merged**